### PR TITLE
(Wallet) Implement Bitcoin support and network selection during onboarding flow

### PR DESCRIPTION
**Pull request by simoarpe**
## Resolves https://github.com/brave/brave-browser/issues/36578

🎨 Figma design: https://www.figma.com/file/oXIiv45pIhF2RGlpEjOfUg/Wallet-onboarding?node-id=6223%3A70942&mode=dev

## UI Preview

|Light Mode |Dark Mode|
|-----|--------|
|Screenshot 2024-04-08 at 3 19 56 PM|Screenshot 2024-04-08 at 3 20 38 PM|

## Demo

https://github.com/brave/brave-core/assets/3308503/70d39c8e-c522-4cd4-b429-50a4e5f6b63a

## Details

* Enable Bitcoin support by default on Android
* Return Bitcoin logo for Bitcoin mainnet and testnet
* Return Neon logo for Neon network 
* Remove unused methods from `CryptoModel`, `NetworkModel`, `NetworkSelectorModel`, and `NetworkSelectorActivity`
* Perform refactoring to onboarding fragments
* Return keyring IDs for Bitcoin mainnet and testnet
* ⚠️ Overhaul viewpager layout, simplify Android layout pass and [fix `notifyDataSetChanged` issue](https://issuetracker.google.com/issues/186366027)
* Implement new `AutoGridLayoutManager` that adapts its columns to recycler view size
* Implement new `OnboardingNetworkSelectionFragment` following Figma design

This PR introduces Bitcoin support for Brave Wallet on Android and implements a new screen that let the user to choose which networks to display. Bitcoin and Filecoin accounts wille be created *only* when their network is selected in the network selection screen. In the network selection screen all the test networks are not selected by default, hence not displayed once the Wallet gets created. Ethereum and Solana accounts instead are pre-selected by default and cannot be deselected. Unselected networks will be added to the hidden network list, accessible in the Wallet settings section (to be implemented soon).

## Follow up Issues

- Add 500 ms delay on Wallet creation screen to avoid flickering - https://github.com/brave/brave-browser/issues/37395
- Refactor fragment creation adapter for onboarding flow - https://github.com/brave/brave-browser/issues/37396
- Implement logic to show hidden network list in the Wallet settings section - https://github.com/brave/brave-browser/issues/37397
- Remove unused code (check also code unused by removed + button) - https://github.com/brave/brave-browser/issues/37398
- Remove live data data type where uneeded - https://github.com/brave/brave-browser/issues/37399
- Fix new network names following Dekstop implementation - https://github.com/brave/brave-browser/issues/37400

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Bitcoin Support Test Plan

- Reset and clear wallet data
- Open Wallet onboarding flow
- On network selection screen, select Bitcoin network and tap continue
- Finish onboarding process
- Navigate to account section
- Observe Bitcoin account is present and is called "Bitcoin Account 1"

### Network Selection Screen Test Plan

- Reset and clear wallet data
- Open Wallet onboarding flow
- On network selection screen, tap on "Show testnets"
- Observe test networks are shown
- Try to deselect Ethereum network
- Observe Ehereum network cannot be deselected
- Try to deselect Solana network
- Observe Solana network cannot be deselected
- Select Bitcoin account
- Observe the button reflects the number of selected networks
- Tap on "Select all" button
- Observe all popular networks get selected
- Tap on search bar
- Type "fil"
- Observe Filecoin networks are shown
- Tap on Filecoin mainnet
- Tap on continue button
- Finish onboarding process
- Observe Bitcoin account and Filecoin account are present

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/onboarding/OnboardingCreatingWalletFragment.java [hunk 1]
```diff
@@ -36,31 +36,50 @@ public View onCreateView(
...
-                        .get(OnboardingViewModel.class);
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        KeyringModel keyringModel = getKeyringModel();
+        if (keyringModel != null) {
+            // Check if a wallet is already present and skip if that's the case.
+            getKeyringModel()
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/onboarding/OnboardingCreatingWalletFragment.java:47]: keyringModel instead getKeyringModel()

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/onboarding/OnboardingCreatingWalletFragment.java [hunk 2]
```diff
@@ -36,31 +36,50 @@ public View onCreateView(
...
+                    .isWalletCreated(
+                            isCreated -> {
+                                if (isCreated) {
+                                    requireActivity().finish();
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/onboarding/OnboardingCreatingWalletFragment.java:51]: we can return here and move `else` block out of `else`

File: android/java/org/chromium/chrome/browser/app/domain/CryptoModel.java [hunk 3]
```diff
@@ -286,10 +268,7 @@ public List<CryptoAccountTypeInfo> getSupportedCryptoAccountTypes() {
 
         @Override
         public List<Integer> getSupportedCryptoCoins() {
-            return getSupportedCryptoAccountTypes()
-                    .stream()
-                    .map(CryptoAccountTypeInfo::getCoinType)
-                    .collect(Collectors.toList());
+            return Arrays.asList(CoinType.ETH, CoinType.SOL, CoinType.FIL, CoinType.BTC);
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/app/domain/CryptoModel.java:271]: Technically there is a feature flag for BTC which might be disabled, so coin should not be seen as supported.

Comment (supermassive) [reply android/java/org/chromium/chrome/browser/app/domain/CryptoModel.java:271]: I would also add that getSupportedCryptoCoins() is only used in NetworkModel to fetch all available networks from core which itself should be refactored to be just getAllNetworks() - not per coin, so getSupportedCryptoCoins would not be needed at all

File: android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java [hunk 4]
```diff
@@ -153,12 +177,175 @@ public AccountInfo getAccount(String address) {
...
+        final Set<NetworkInfo> addHiddenNetworks = new HashSet<>();
+
+        MutableLiveData<Integer> removeHiddenNetworksLiveData = new MutableLiveData<>();
+        MutableLiveData<Integer> addHiddenNetworksLiveData = new MutableLiveData<>();
+
+        AtomicInteger countRemovedHiddenNetworks = new AtomicInteger(0);
+        AtomicInteger countAddedHiddenNetworks = new AtomicInteger(0);
+
+        AtomicBoolean removeHiddenNetworksDone = new AtomicBoolean(false);
+        AtomicBoolean addHiddenNetworksDone = new AtomicBoolean(false);
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:216]: why we need AtomicBoolean? Everything in this file should be main-thread only

File: android/java/org/chromium/chrome/browser/app/domain/CryptoModel.java (hunk 3 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/domain/CryptoModel.java:271]: Yes! It's something I would like to remove soon, but the PR was already fairly big, so to keep scope smaller I've created a follow-up issue for this 👉 https://github.com/brave/brave-browser/issues/37398

File: android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java [hunk 5]
```diff
@@ -113,31 +120,48 @@ public void getAccounts(Callbacks.Callback1<AccountInfo[]> callback1) {
...
-        mKeyringService.addAccount(coinType, keyringId, accountName,
-                result -> { handleAddAccountResult(result, callback); });
+        mKeyringService.addAccount(
+                coinType,
+                keyringId,
+                accountName,
+                result -> {
+                    handleAddAccountResult(result, callback);
+                });
     }
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:140]: This is just change in formatting. Do we have new formatting style? Or it was not properly formatted before? or?

Comment (supermassive) [reply android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:140]: Just curious

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/AccountSelectorActivity.java [hunk 6]
```diff
@@ -26,12 +27,16 @@
...
+import java.util.Arrays;
 import java.util.List;
 
 public class AccountSelectorActivity extends BraveWalletBaseActivity
         implements AccountSelectorItemListener {
     private static final String TAG = "AccountSelector";
 
+    private static final List<Integer> SUPPORTED_COIN_TYPES =
+            Arrays.asList(CoinType.ETH, CoinType.SOL, CoinType.FIL);
+
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/AccountSelectorActivity.java:39]: why no BTC?

File: android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java (hunk 5 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:140]: Very odd, right? 
The only explaination I can give is that we have a slightly different formatting style. I didn't pay much attention to this as it's done automatically, but I'll investigate and get back to you.

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/AccountSelectorActivity.java (hunk 6 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/AccountSelectorActivity.java:39]: It's for DApps, and as far as I know we don't support Bitcoin (yet). I'm going to rename the constant into something more meaninful like `SUPPORTED_COIN_TYPES_ON_DAPP`.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java [hunk 7]
```diff
@@ -865,6 +872,9 @@ public static String getNetworkIconName(String chainId, @CoinType.EnumType int c
                 case CoinType.FIL:
                     logo = "fil.png";
                     break;
+                case CoinType.BTC:
+                    logo = "btc.png";
+                    break;
                 default: // Do nothing
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java:878]: there is no localhost for btc

Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java:878]: Thanks, I'll remove it!

File: android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java (hunk 5 above)
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:140]: iirc we applied Java formatting ~2 years ago, most likely that file was created before that date. We have it pretty often in old files.

File: browser/ui/android/strings/android_brave_strings.grd [hunk 8]
```diff
@@ -3065,6 +3089,9 @@ If you don't accept this request, VPN will not reconnect and your internet conne
     <message name="IDS_BRAVE_WALLET_CREATE_ACCOUNT_FILECOIN_DESCRIPTION" desc="Add account modal create account Filecoin description">
       Store FIL asset
     </message>
+    <!-- <message name="IDS_BRAVE_WALLET_CREATE_ACCOUNT_BITCOIN_DESCRIPTION" desc="Add account modal create account Bitcoin description">
```
[the remark this answers is not in the document]
Comment (simoarpe) [reply browser/ui/android/strings/android_brave_strings.grd:3092]: My bad, they are uneeded strings that must have slipped into the commit. I'm removing these now.

**Review by mkarolin — APPROVED** [review]
`strings`++

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java (hunk 5 above)
Comment (supermassive) [reply android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:141]: That would be great if such formatting rules are forcedly applied to all .java files(maybe iteratively per component in many PRs), so reviewers would not have to review what they don't need to review.

**Review by supermassive — APPROVED** [review]

File: android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java (hunk 5 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:141]: I've created an issue for Wallet files: https://github.com/brave/brave-browser/issues/37454

**Pull request merged**